File: js/graph.js

//#region DATASETS
const vertices = ["A", "B", "C", "D", "E"];

const vertexIdxs = {
    "A": 0,
    "B": 1, 
    "C": 2, 
    "D": 3, 
    "E": 4
}
const edges = [
    ["A", "B"],
    ["A", "D"],
    ["B", "C"],
    ["C", "D"],
    ["C", "E"],
    ["D", "E"]
];

const adjacencyMatrix = [
    [0, 1, 0, 1, 0],
    [1, 0, 1, 0, 0],
    [0, 1, 0, 1, 1],
    [1, 0, 1, 0, 1],
    [0, 0, 1, 1, 0]
];

const adjencyList = [
    ["B", "D"],
    ["A", "C"],
    ["B", "D", "E"],
    ["C", "D"],
    ["A", "C", "E"],
    ["C", "D"]
]




//#endregion
class Node {
    constructor(value) {
        this.value = value;
        this.edgesList = []; 
    }

    connect(node) {
        this.edgesList.push(node);
        node.edgesList.push(this);
    }

    getAdjacentNodes() {
        return this.edgesList.map(n => n.value);
    }

    isConnected(node) {
        return this.edgesList.some((n) => {
            return node.value === n.value;
        })
    }
}

class Graph {
    constructor(nodes) {
        this.nodes = [...nodes]
    }
    addToGraph(node) {
        this.nodes.push(node);
    }

    breadthFirstTraversal(start, end) {
        const queue = [start];
    
        const visited = new Set();
        visited.add(start);
        while(queue.length > 0) {
            const node = queue.shift();

            if (node.value === end.value) {
                console.log("Found it!");
                return;
            }

            for (const edge of node.edgesList) {
                if (!visited.has(edge)) {
                    queue.push(edge);
                    visited.add(edge);
                }
            }

            console.log(node.value)
        }
    }

    findShortestPath(start, end) {
        const queue = [start];
        const visitedNodes = {};
        visitedNodes[start.value] = null;

        while (queue.length > 0) {
            const node = queue.shift();
            if (node.value === end.value){
               return this.reconstructPath(visitedNodes, start, end);
   
            }
            

            for (const adjacency of node.edgesList) {
                if (!visitedNodes.hasOwnProperty(adjacency.value)) {
                    visitedNodes[adjacency.value] = node;
                    queue.push(adjacency);
                    console.log(adjacency.value)
                }
            }
        }
       
    }

    reconstructPath(visitedNodes, startNode, endNode) {
        let currNode = endNode;
        let path = [];
        while(currNode !== null) {
            let prev = currNode.value;
            path.unshift(prev)
            currNode = visitedNodes[prev];
        }

        return path.join("-->");
        
    }

    depthFirstSearch(start, end, visited) {
        if (!visited) {
            visited = new Set();
        }
        if (start.value === end.value) {
            console.log("Destination reached at : ", end.value);
            return;
        }
        visited.add(start)
       
        console.log("Visiting -->", start.value)

        for (let adjacency of start.edgesList) {
            if (!visited.has(adjacency)) {
                visited.add(adjacency);
                this.depthFirstSearch(adjacency, end, visited);
            }
        }
        
            

    }

}

//#region EDGES

const findAdjacent = (node) => {
    let adjacents = [];
   for (let edge of edges) {
        const nodeIdx = edge.indexOf(node);
        if (nodeIdx > -1) {
            const adjNode= nodeIdx === 0 ? edge[1]: edge[0];
            adjacents.push(adjNode);
        }
   }

   return adjacents
}

// console.log(findAdjacent("A"));

const isConnected = (node1, node2) => {
    return edges.some((edge) => {
        return edge.indexOf(node1) > -1 && edge.indexOf(node2) > -1;
    })
}

// console.log(isConnected("C", "D"));
// console.log(isConnected("A", "E"));

//#endregion

//#region Matrix
const findAdjenciesMatrix = (node) => {
    const adjacents = [];
    const nodeIdx = vertexIdxs[node];

    adjacencyMatrix[nodeIdx].forEach((elt, index) => {
        if (elt === 1) {
            adjacents.push(vertices[index]);
        }
    })
    return adjacents;
}

// console.log(findAdjenciesMatrix("A"));
// console.log(findAdjenciesMatrix("C"));

const isConnectedMatrix = (node1, node2) => {
    const node1Idx = vertexIdxs[node1];
    const node2Idx = vertexIdxs[node2];
    return !!adjacencyMatrix[node1Idx][node2Idx];
   
}
// console.log(isConnectedMatrix("A", "B"));
// console.log(isConnectedMatrix("A", "E"));

//#endregion

//#region List

const nodeA = new Node('A');
const nodeB = new Node('B');
const nodeC = new Node('C');
const nodeD = new Node('D');
const nodeE = new Node('E');

nodeA.connect(nodeB);
nodeA.connect(nodeD);
nodeB.connect(nodeC);
nodeC.connect(nodeD);
nodeC.connect(nodeE);
nodeD.connect(nodeE);


const graph = new Graph([
    nodeA,
    nodeB,
    nodeC,
    nodeD,
    nodeE
]);

// for (let node of graph.nodes) {
//     console.log(node.value);
//     for (connectedNode of node.edgesList) {
//         console.log(`Node ${node.value} is connected to ${connectedNode.value}`);
//     }
// }

// console.log(nodeA.isConnected(nodeE));
//#endregion

//#region Breadth First and Depth first

const DFW = new Node('DFW');
const JFK = new Node('JFK');
const LAX = new Node('LAX');
const HNL = new Node('HNL');
const SAN = new Node('SAN');
const EWR = new Node('EWR');
const BOS = new Node('BOS');
const MIA = new Node('MIA');
const MCO = new Node('MCO');
const PBI = new Node('PBI');

const airportsGraph = new Graph([
    DFW,
    JFK,
    LAX,
    HNL,
    SAN,
    EWR,
    BOS,
    MIA,
    MCO,
    PBI
]);

DFW.connect(LAX);
DFW.connect(JFK);
LAX.connect(HNL);
LAX.connect(EWR);
LAX.connect(SAN);
JFK.connect(BOS);
JFK.connect(MIA);
MIA.connect(MCO);
MIA.connect(PBI);
MCO.connect(PBI);

// console.log(airportsGraph.findShortestPath(DFW, MIA));
// airportsGraph.depthFirstSearch(DFW, MCO)

// Topological order

//#endregion

File: golang/datastructs/graph.go
package datastructs

import (
	"fmt"
	"strconv"
)

type value interface {
	string | int
}

type arr[T value] []GraphNode[T]

func (a *arr[T]) shift() GraphNode[T] {
	elt := (*a)[0]
	*a = (*a)[1:]
	return elt
}

func (a *arr[T]) pop() GraphNode[T] {
	i := len((*a)) - 1
	elt := (*a)[i]
	*a = (*a)[0:i]
	return elt
}

/***********************************
*          	  NODE
***********************************/

type GraphNode[T value] struct {
	value     T
	edgesList []*GraphNode[T]
}

func newNode[T value](val T) GraphNode[T] {
	return GraphNode[T]{value: val, edgesList: []*GraphNode[T]{}}
}

func (n *GraphNode[T])Connect(new *GraphNode[T]) {
	n.edgesList = append(n.edgesList, new)
	new.edgesList = append(new.edgesList, n)
}

func (n *GraphNode[T]) findAdjacents() {
	adj := []T{}
	for _, v := range n.edgesList {
		adj = append(adj, v.value)
	}
	fmt.Println(adj)
}

func (n *GraphNode[T]) printEdges() []T {
	edges := []T{}
	for _, v := range n.edgesList {
		edges = append(edges, v.value)
	}
	return edges
}

/***********************************
*          	 GRAPH
***********************************/
type graph[T value] struct {
	nodes []GraphNode[T]
}

func (g *graph[T]) print() {
	for _, v := range g.nodes {
		edges := v.printEdges()
		fmt.Println(v.value, edges)
	}
}

func (g *graph[T]) add(node GraphNode[T]) {
	g.nodes = append(g.nodes, node)
}

func (g *graph[T]) sum() int {
	sum := 0
	for _, node := range g.nodes {
		num, err := strconv.Atoi(fmt.Sprintf("%v", node.value))
		if err != nil {
			fmt.Println("Wrong type")
		}
		sum += num
	}
	return sum
}

func (g *graph[T]) breadthFirstPrint() {
	q := arr[T]{g.nodes[0]}
	visited := map[T]bool{
		g.nodes[0].value: true,
	}

	for len(q) > 0 {
		n := q.shift()
		for _, e := range n.edgesList {
			if !visited[e.value] {
				q = append(q, *e)

				visited[e.value] = true
			}

		}
		fmt.Println(n.value)
	}

}

func recursiveBreadthFirst[T value](node GraphNode[T], visited map[T]bool) {
	if visited == nil {
		visited = map[T]bool{}
	}

	visited[node.value] = true

	fmt.Println("visiting-->", node.value)
	for _, edge := range node.edgesList {
		if !visited[edge.value] {

			visited[edge.value] = true
			recursiveBreadthFirst(*edge, visited)
		}
	}
}

func (g *graph[T]) depthFirstPrint() {
	stack := arr[T]{g.nodes[0]}
	visited := map[T]bool{
		g.nodes[0].value: true,
	}

	for len(stack) > 0 {
		node := stack.pop()
		for _, edge := range node.edgesList {
			if !visited[edge.value] {
				stack = append(stack, *edge)
				visited[edge.value] = true
			}
		}
		fmt.Println(node.value)
	}

}

func shortestPath[T value](start, end GraphNode[T]) {
	queue := arr[T]{start}
	visited := map[T]*T{
		start.value: nil,
	}

	for len(queue) > 0 {
		node := queue.shift()
		if node.value == end.value {
			reconstructPath(visited, end)
			return
		}

		for _, edge := range node.edgesList {
			if _, ok := visited[edge.value]; !ok {
				visited[edge.value] = &node.value
				queue = append(queue, *edge)
			}
		}
	}

}

func reconstructPath[T value](nodeList map[T]*T, endNode GraphNode[T]) {
	current := &endNode.value
	pathStr := ""
	for current != nil {
		previous := current
		if pathStr == "" {
			pathStr = fmt.Sprintf("%v", *previous)
		} else {
			pathStr = fmt.Sprintf("%v", *previous) + "-->" + pathStr
		}
		current = nodeList[*previous]

	}
	fmt.Println(pathStr)
}

func graphPlay() {
	gi := graph[int]{
		nodes: []GraphNode[int]{},
	}

	nodeA := newNode(8)
	nodeB := newNode(34)
	nodeC := newNode(25)
	nodeD := newNode(5)
	nodeE := newNode(12)

	nodeA.Connect(&nodeB)
	nodeA.Connect(&nodeD)
	nodeB.Connect(&nodeC)
	nodeC.Connect(&nodeD)
	nodeD.Connect(&nodeE)

	gi.add(nodeA)
	gi.add(nodeB)
	gi.add(nodeC)
	gi.add(nodeD)
	gi.add(nodeE)

	// fmt.Println(gi.sum())
	// nodeA.findAdjacents()

	DFW := newNode("DFW")
	JFK := newNode("JFK")
	LAX := newNode("LAX")
	HNL := newNode("HNL")
	SAN := newNode("SAN")
	EWR := newNode("EWR")
	BOS := newNode("BOS")
	MIA := newNode("MIA")
	MCO := newNode("MCO")
	PBI := newNode("PBI")

	DFW.Connect(&LAX)
	DFW.Connect(&JFK)
	LAX.Connect(&HNL)
	LAX.Connect(&EWR)
	LAX.Connect(&SAN)
	JFK.Connect(&BOS)
	JFK.Connect(&MIA)
	MIA.Connect(&MCO)
	MIA.Connect(&PBI)
	MCO.Connect(&PBI)

	gs := graph[string]{
		nodes: []GraphNode[string]{
			DFW,
			JFK,
			LAX,
			HNL,
			SAN,
			EWR,
			BOS,
			MIA,
			MCO,
			PBI,
		},
	}

	gs.print()
	// gs.depthFirstPrint()
	// recursiveBreadthFirst(DFW, nil)
	shortestPath(DFW, MIA)
}
